src/java/org/apache/cassandra/db/tries/VerificationCursor.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.tries;
+
+import java.util.Arrays;
+
+import com.google.common.base.Preconditions;
+
+import org.agrona.DirectBuffer;
+import org.apache.cassandra.utils.Hex;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+
+public interface VerificationCursor
+{
+    int EXHAUSTED_DEPTH = -1;
+    int EXHAUSTED_TRANSITION = -1;
+    int INITIAL_TRANSITION = -1;
+
+    /**
+     * Verifies:
+     * - advance does advance, depth <= prevDepth + 1 and transition is higher than previous at the same depth
+     *   (this requires path tracking)
+     * - skipTo is not called with earlier or equal position (including lower levels)
+     * - maybeSkipTo is not called with earlier position that can't be identified with depth/incomingTransition only
+     *   (i.e. seeks to lower depth with an incoming transition that lower than the previous at that depth)
+     * - exhausted state is depth = -1, incomingTransition = -1 (maybe change to 0?)
+     * - start state is depth = 0, incomingTransition = -1 (maybe change to 0?)
+     */
+    class Plain<T, C extends Cursor<T>> implements Cursor<T>, Cursor.TransitionsReceiver
+    {
+        final Direction direction;
+        final C source;
+        final int minDepth;
+        int returnedDepth;
+        int returnedTransition;
+        byte[] path;
+
+        Cursor.TransitionsReceiver chainedReceiver = null;
+        boolean advanceMultipleCalledReceiver;
+
```

Comment:
Nit: it's not a receiver but just a boolean flag, a simpler name `advanceMultipleCalled` would be a bit more readable